packages/inngest/src/components/InngestCommHandler.ts

```diff
@@ -677,31 +677,102 @@ export class InngestCommHandler<
           },
         });
 
+      const assumedMode = getMode({ env: this.env, client: this.client });
+
+      if (assumedMode.isExplicit) {
+        this._mode = assumedMode;
+      } else {
+        const serveIsProd = await actions.isProduction?.(
+          "starting to handle request"
+        );
+        if (typeof serveIsProd === "boolean") {
+          this._mode = new Mode({
+            type: serveIsProd ? "cloud" : "dev",
+            isExplicit: false,
+          });
+        } else {
+          this._mode = assumedMode;
+        }
+      }
+
+      this.upsertKeysFromEnv();
```

Comment:
Fetching modes (and late actions like using environment variables passed in the request) is moved much earlier, so we can appropriately validate this early too.